File: Trees/deepestLeavesSum.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */

/*
An emulation of the list comprehension methodn used in my python solution

Starting with the a queue that just has the root in it
Use n to keep track of all the nodes at that level
Iterate through all the nodes at their level, adding their children to the next q
If the node doesn't have children you know it's a leaf, so you add its value to sum
You only return this value of sum, however, when you none of the leaves at a level have children
	(i.e. they are the deepest leaves)
Hence the deepest leaves sum is returned

Time complexity 	O(n)
Space complexity	O(log n)
*/

class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        queue <TreeNode*> q;
        q.push(root);
        int sum = 0;
        while (!q.empty()){
            int n = q.size();
            sum = 0;
            while (n--){
                cout << sum << endl;
                TreeNode* node = q.front(); q.pop();
                if (node->left){
                    q.push(node->left);
                }
                if (node->right){
                    q.push(node->right);
                }
                if(!node->left && !node->left){
                    sum += node->val;
                }
            }
        }
        return sum;
    }
};

/*
//This a tidier version of my logic

class Solution {
public:
    int deepestLeavesSum(TreeNode* root) {
        int res = 0, i;
        queue<TreeNode*> q;
        q.push(root);
        while (q.size()) {
            for (i = q.size() - 1, res = 0; i >= 0; --i) { //Reinitialise res to 0
                TreeNode* node = q.front(); q.pop();
                res += node->val;
                if (node->right) q.push(node->right);
                if (node->left)  q.push(node->left);
            }
        }
        return res;
    }
};
*/

File: Trees/deepestLeavesSum.py

'''
https://leetcode.com/problems/deepest-leaves-sum

This algorithm is basically a level order traversal
but we have saved in memory both the current and previous levels

Add root to queue
While there are nodes in the queue
   Using list comprehension set the previous level to the current nodes in the queue
   At the same time set the queue equal to the children of the current nodes in the queue (provided they exist)
Once the queue is empty, your at the last level. Sum all the node values

Time complexity     O(n)
Space complexity    O(n)
'''

class Solution:
    def deepestLeavesSum(self, root: TreeNode) -> int:
        queue = [root]
        while queue:
            pre, queue = queue , [child for node in queue for child in [node.left, node.right] if child]
        return sum(leaf.val for leaf in pre)


#Depth first search
#Keep track of depth
#If depth less than max, no need to add to sum
#If depth equal max depth add to sum
#If depth greater than max set sum to that value
#Return sum

class Solution:
    def deepestLeavesSum(self, root: TreeNode) -> int:
        depth = sum = maxDepth = 0
        return traverse(root, depth, sum, maxDepth)

def traverse(Node, depth, sum, maxDepth):
    if (Node):
        print(Node.val, depth, maxDepth)
        depth += 1
        traverse(Node.left, depth, sum, maxDepth)
        traverse(Node.right, depth, sum, maxDepth)
        if (depth == maxDepth):
            sum += Node.val
        elif (depth > maxDepth):
            sum = Node.val
            maxDepth = depth
        #print(Node.val, depth)
        depth -= 1
    return sum
